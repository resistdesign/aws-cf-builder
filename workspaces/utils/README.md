# Utils

Create reusable resource packs and add common assets to your CloudFormation templates.

## Install

- NPM: `npm i -S @aws-cf-builder/utils`
- Yarn: `yarn add @aws-cf-builder/utils`

## Usage

See the [example](../example/README.md) project.
