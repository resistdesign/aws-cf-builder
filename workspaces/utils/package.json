{
  "name": "@aws-cf-builder/utils",
  "version": "1.4.0",
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "author": "Resist Design",
  "license": "MIT",
  "scripts": {
    "build": "tsc --outDir dist"
  },
  "prettier": {
    "printWidth": 150,
    "semi": true,
    "singleQuote": true,
    "trailingComma": "es5",
    "useTabs": false
  },
  "devDependencies": {
    "prettier": "^2.3.2"
  },
  "dependencies": {
    "@aws-cf-builder/types": "^1.4.0",
    "@types/flat": "^5.0.2",
    "flat": "^5.0.2",
    "tslib": "^2.3.0",
    "typescript": "^4.3.5"
  }
}
