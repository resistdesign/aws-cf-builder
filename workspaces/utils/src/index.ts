import { CloudFormationParameter, CloudFormationTemplate } from '@aws-cf-builder/types';

export type ParameterInfo = {
  ParameterId: string;
  Parameter: CloudFormationParameter;
  Label: string;
  Group?: string;
};

export const addParameter = (parameterInfo: ParameterInfo, template: CloudFormationTemplate): CloudFormationTemplate => {
  const { ParameterId, Parameter, Label, Group } = parameterInfo;
  const { Parameters, Metadata: { 'AWS::CloudFormation::Interface': { ParameterGroups = [], ParameterLabels = {} } = {} } = {} } = template;

  let NewParameterGroups = ParameterGroups;

  if (Group) {
    const GroupObject = ParameterGroups.filter((g) => g.Label === Group)[0];

    NewParameterGroups = GroupObject
      ? ParameterGroups.map((g) =>
          g.Label === Group
            ? {
                ...g,
                Parameters: [...(g.Parameters || []), ParameterId],
              }
            : g
        )
      : [
          {
            Label: {
              default: Group,
            },
            Parameters: [ParameterId],
          },
        ];
  }

  return {
    ...template,
    Parameters: {
      ...Parameters,
      [ParameterId]: Parameter,
    },
    Metadata: {
      ...template.Metadata,
      'AWS::CloudFormation::Interface': {
        ...template?.Metadata?.['AWS::CloudFormation::Interface'],
        ParameterGroups: NewParameterGroups,
        ParameterLabels: {
          ...ParameterLabels,
          [ParameterId]: {
            default: Label,
          },
        },
      },
    },
  };
};
