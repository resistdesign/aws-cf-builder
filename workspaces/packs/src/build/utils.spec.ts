import { COMMAND_HELPERS, createBuildSpec } from './utils';

describe('Build Utils', () => {
  describe('createBuildSpec', () => {
    test('should create a YAML build spec', () => {
      const buildSpec = createBuildSpec({
        phases: {
          install: {
            'run-as': 'main-user',
            commands: [COMMAND_HELPERS.addNPMTokenWithNPMRC({ npmToken: 'lkj5thlk5jh345lkj5h54jkth54k2l' })],
          },
          build: {
            'run-as': 'main-user',
            'on-failure': 'ABORT',
            commands: [
              COMMAND_HELPERS.copyDirectoryToS3({ s3Domain: 'example.com', directoryPath: './dist' }),
              COMMAND_HELPERS.updateFunction({
                cloudFunctionArn: 'th5ljk2hl5kj4h325',
                codeZipFilePath: './dist/code.zip',
              }),
            ],
          },
          post_build: {
            'run-as': 'main-user',
            commands: [COMMAND_HELPERS.cloudFrontInvalidation({ cloudFrontDistributionId: 'CFDID' })],
          },
        },
      });

      expect(buildSpec).toMatchSnapshot();
    });
  });
});
