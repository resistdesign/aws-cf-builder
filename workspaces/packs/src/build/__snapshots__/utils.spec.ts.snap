// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build Utils createBuildSpec should create a YAML build spec 1`] = `
"version: 0.2
phases:
  install:
    run-as: main-user
    commands:
      - echo '//registry.npmjs.org/:_authToken=lkj5thlk5jh345lkj5h54jkth54k2l' >
        .npmrc
  build:
    run-as: main-user
    on-failure: ABORT
    commands:
      - aws s3 cp --recursive --acl public-read ./dist s3://example.com/
      - aws lambda update-function-code --function-name \\"th5ljk2hl5kj4h325\\"
        --zip-file \\"fileb://./dist/code.zip\\"
  post_build:
    run-as: main-user
    commands:
      - aws cloudfront create-invalidation --distribution-id \\"CFDID\\" --paths \\"/*\\"
"
`;
